### PR TITLE
Write tests for datetime queries

### PR DESCRIPTION
**Pull request by manzoorwanijk**
Cmpletes #2097 

## Checklist

* [ ] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [ ] User input is adequately validated and sanitized
* [ ] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [ ] My code follows the Event Espresso code style.
* [ ] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

File: assets/prototype/domain/eventEditor/context/ContextProviders.js [hunk 1]
```diff
@@ -1,28 +1,41 @@
...
+
 /**
  * A collection of top level providers that are used by multiple parts of the application.
+ * This is also shared by unit tests.
  *
  * @param {ReactElement} children The element that should be wrapped.
  * @returns {ReactElement} The wrapped element.
  */
-const ContextProviders = ({ children }) => (
+export const CommonProviders = ({ children }) => (
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/context/ContextProviders.js]: :+1:

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/contextWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import { MockedProvider, MockedResponse } from '@apollo/react-testing';
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/datetimes/test/contextWrapper.tsx]: Yes, very cool:
```
The MockedProvider is a test-utility that allows you to create a mocked version of the ApolloProvider that doesn't send out network requests to your API, but rather allows you to specify the exact response payload for a given request.
```

Comment (nicolad) [reply assets/prototype/domain/eventEditor/data/queries/datetimes/test/contextWrapper.tsx]: And `MockedResponse` is also nice interface: https://github.com/apollographql/react-apollo/blob/master/packages/testing/src/mocks/types.ts

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	variables: {
+		where: {
+			eventId,
+		},
+	},
+};
+
+export const setup = () => {
+	window.eeEditorEventData = { eventId };
+};
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts]: We might need a test utility folder for this kind of stuff.

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		reserved: 0,
+		sold: 6,
+		startDate: 'end data',
+		__typename: 'EspressoDatetime',
+	},
+];
+
+export const edge = {
+	nodes,
+	__typename: 'EspressoRootQueryDatetimesConnection',
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts]: Now I see why you asked about those generated types :)
https://github.com/eventespresso/event-espresso-core/blob/EDTR/master/assets/prototype/domain/shared/types.d.ts#L2203

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import { useEffect } from 'react';
+import { useApolloClient } from '@apollo/react-hooks';
+import { renderHook } from '@testing-library/react-hooks';
+
+import useDatetimes from '../useDatetimes';
+import contextWrapper from './contextWrapper';
+import { successMocks, setup, cleanup, nodes, request, edge } from './data';
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx]: Nice extraction :+1:

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import { useStatus, TypeName } from '../../../../../../application/services/apollo/status';
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx]: Not directly related to this PR, but we need to do something with these `../../../../../../` :smiley:

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+				// make sure the status flags is set for useDatetimes to work
+				useEffect(() => {
+					setIsLoaded(TypeName.datetimes, true);
+				}, []);
+				return useDatetimes();
+			},
+			{ wrapper }
+		);
+
+		expect(result.current.length).toBe(nodes.length);
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx]: maybe extract `result.current` into some variable, like `cachedDates` or similar

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts (hunk 3 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts]: Yes, that's there for time being.

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts (hunk 4 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts]: No, that wasn't the actual reason. I actually tried to create the test data dynamically but that didn't fit well and was difficult for others to understand.

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx (hunk 6 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx]: We can convert those hooks to a package.

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx (hunk 7 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx]: :+1:

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx (hunk 6 above)
Comment (nicolad) [reply assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.tsx]: Yes, good suggestion.
We can do monorepo as in GB or as in this example: https://github.com/jannikbuschke/lerna-react-typescript-sample

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts (hunk 3 above)
Comment (nicolad) [reply assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts]: :man_facepalming: 
sorry
:smiley:

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts [hunk 8]
```diff
@@ -1,11 +1,13 @@
 import { GraphQLError } from 'graphql';
 
 import '../../../../types';
+import { Datetime, DatetimeEdge } from '../../../types';
+import { ReadQueryOptions } from '../../types';
 import { GET_DATETIMES } from '../';
 
-const eventId = 100;
+export const eventId: number = 100;
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/datetimes/test/data.ts]: That is ultra safety :smiley:

File: assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+				// make sure the status flags is set for useDatetimes to work
+				useEffect(() => {
+					setIsLoaded(TypeName.datetimes, true);
+				}, []);
+				return useDatetimes();
+			},
+			{ wrapper }
+		);
+
+		const { current: cachedDatetimes } = result;
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/data/queries/datetimes/test/useDatetimes.test.ts]: Thx :bowing_man:

**Pull request merged**